### PR TITLE
Updated Upstream (Bukkit/CraftBukkit/Spigot)

### PR DESCRIPTION
**Pull request by Machine-Maker**
Updated Upstream (Bukkit/CraftBukkit/Spigot)

Upstream has released updates that appear to apply and compile correctly.
This update has not been tested by PaperMC and as with ANY update, please do your own testing

Bukkit Changes:
7b1856bf PR-278: Add API to check and set if horses are grazing hay
0111185d PR-275: Add isJohnny / setJohnny for Vindicator
ea81107b PR-274: Add trap related methods to SkeletonHorse
eeda7998 Make EnderDragon extend/implement Mob
96509e04 PR-702: fromNodeTree is now private, this does not belong to the API
ead93bd8 PR-701: Option to configure the max width of a line in the config
becb47c4 PR-699: Add missing information to javadoc of ItemFlag#HIDE_POTION_EFFECTS
f3a8a4ce SPIGOT-5766: Updates related to the villager trading changes introduced in MC 1.14.
79cb3d7f PR-691: Add Material#getDefaultAttributeModifiers
65ff9376 SPIGOT-6866: Fixed spaces a between comment indicator and actual comment
8a747588 Restore protected members to YamlConfiguration
ca67e0f3 PR-693: Add proper javadocs for dolphins grace, bad omen, and hero of the village
3e2dd2bc SPIGOT-3247: Comment support for YAML files

CraftBukkit Changes:
c93fcba4 PR-378: Add API to check and set if horses are grazing hay
752fa511 PR-374: Add isJohnny / setJohnny for Vindicator
19f2af06 PR-373: Add trap related methods to SkeletonHorse
b8a25cbd Fix map colour not being marked as a handled tag
eea8d559 Make EnderDragon extend/implement Mob
633ac8b3 Add self-riding check to Entity#addPassenger
0ba6b901 SPIGOT-5766: The VillagerReplenishTradeEvent is now called when a villager restocks its trades.
bdff7d56 SPIGOT-5627: Leads vanish when transporting leashed entities through a Nether portal
319ca703 PR-978: Add Material#getDefaultAttributeModifiers
c59b0ca1 SPIGOT-6859: PlayerInteractEntityEvent#setCancelled desync with Axolotls

Spigot Changes:
78a14d9d SPIGOT-6644: Previously inactive experience orbs may not target players

### CODE REVIEW

**Review by kennytv — DISMISSED** [review]
Spigot broke some config stuff, we'll either have to wait or fix them ourselves if there's more important commits to upstream

**Comment by HexedHero** [comment]
> Spigot broke some config stuff, we'll either have to wait or fix them ourselves if there's more important commits to upstream

Isn't this like the 3rd time Spigot has tried to update SnakeYAML to support comments?

**Comment by MiniDigger** [comment]
https://hub.spigotmc.org/jira/browse/SPIGOT-6865 for example

**Comment by LlmDl** [comment]
I'd really appreciate it if this wasn't merged right away.

**Comment by Phoenix616** [comment]
Pretty sure this is pretty stable now, haven't heard of a single plugin using the config system properly which breaks. (And the author of the SnakeYAML and the Spigot PR is the dev of BetonQuest which heavily relies on config saving with comments and was the reason behind implementing this change so it has been tested against that)

**Comment by Wolf2323** [comment]
So, now i also have to write something here, as i am the person who invested now nearly a year, to develop the comment support.

I am sure, that the implementation and the usual use of configs works as before. The are minor "api" changes that break some CUSTOM stuff, but the point of this changes are that they where never intendet to be real public API, or it was a neccecery change, to support comment.

The main "broken" thing at the momment, its the "saveToString" method. Before it did not contain comments, and now it do. But that is selfexplaining. For anyone who run in this problem, but hwo do not want to simply drop there custom implementation, simply set parseComments to false in the FileConfigurationOptions.

I recomment to simply wait a bit, until things has been relext, i dont think that there are major issues, but lets wait a bit.

There are ruffently this related issues: 
- https://hub.spigotmc.org/jira/browse/SPIGOT-3247 
  General information, no issue
- https://hub.spigotmc.org/jira/browse/SPIGOT-6865
   custom implementation
- https://hub.spigotmc.org/jira/browse/SPIGOT-6866
   solved
- https://hub.spigotmc.org/jira/browse/SPIGOT-6867
  discussion about the case, when header comments are may duplicated, bad api use
- https://bitbucket.org/snakeyaml/snakeyaml/issues/518/comments-could-cause-queue-full
  to many comments cause a full queue in snakeyaml
  PR for fix: https://bitbucket.org/snakeyaml/snakeyaml/pull-requests/3

**Comment by Wolf2323** [comment]
keep a look at the last comment from me, there are now all issues listed https://hub.spigotmc.org/jira/browse/SPIGOT-3247

**Comment by zml2008** [comment]
Whether intended or not, `YamlConfiguration` has been non-final since October 2011 when it was added, and there *will* be plugins that depend on this behavior.

From working with SnakeYAML's comment handling code, I think it's likely that for any content emitted by Bukkit's configuration API, the result will be successfully loadable with comment parsing enabled. However, there are various YAML constructs that still have issues (I know commented flow mappings don't parse correctly, and I would not be surprised if there are more), so it is possible some user-edited configurations will fail to load with these changes.

I think a reasonable compromise would be to either default comment parsing to off (would have to be on the spigot side), or provide a system property that controls the default comment handling state, to mitigate potential issues with unmaintained plugins or work around bugs in the field. This is what I've [done in my work on Configurate](https://github.com/SpongePowered/Configurate/blob/9d4359647e3d76c88ca58c905293b43216231597/format/yaml/src/main/java/org/spongepowered/configurate/yaml/YamlConfigurationLoader.java#L162).

Given one of those escape hatches and a few weeks for the dust to settle from plugin authors reacting to Spigot changes, I think comment handling should be fine to merge.

**Comment by Wolf2323** [comment]
@zml2008 as i not only developed the Spigot/Bukkit implementation of comments, but also the SnakeYaml side, i would be really happy, if you can provide a failing example, if that is the case, i will imidently fix it.

**Comment by zml2008** [comment]
See this patch: https://github.com/SpongePowered/Configurate/blob/6bbe06f4ec37108c7d67ba7ecdd8ac01af7d7176/vendor/snakeyaml-patches/0001-parser-More-tests-for-comment-reading.patch#L221 (yes, some of the patch is likely not relevant anymore due to your flow sequence fix, but I haven't had a chance to spend much time on Configurate recently)

the relevant test added to `ParserWithCommentsEnabledTest`:

```java
    @Test
    public void testCommentsInFlowMapping() {
        String data = "" +
            "# Beginning\n" +
            "{\n" +
            "    # Hello\n" +
            "    key1: meow,\n" +
            "    # World\n" +
            "    key2: purr\n" +
            "}\n";

        List<ID> expectedEvents = Arrays.asList(
            ID.StreamStart,
            ID.Comment,
            ID.DocumentStart,
            ID.MappingStart,
            ID.Comment,
            ID.Scalar, ID.Scalar,
            ID.Comment,
            ID.Scalar, ID.Scalar,
            ID.MappingEnd,
            ID.DocumentEnd,
            ID.StreamEnd
        );

        final Scanner scanner = new ScannerImpl(new StreamReader(data)).setParseComments(true);
        do {
            System.out.println(scanner.getToken().getTokenId().name());
        } while (!scanner.checkToken(Token.ID.StreamEnd));

        final Parser parser = createParser(data);
        assertEventListEquals(expectedEvents, parser);
    }
```

**Comment by Wolf2323** [comment]
@zml2008 i fixed it in this PR: https://bitbucket.org/snakeyaml/snakeyaml/pull-requests/2

people should stop fixing stuff in there on implementation, and instead open issues in the related project. So every one get a fix, and people like i am know that there is an issue.

File: patches/api/0352-Add-system-prop-for-disabling-config-comment-parsing.patch [hunk 1]
```diff
@@ -0,0 +1,49 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jake Potrebic <jake.m.potrebic@gmail.com>
+Date: Thu, 30 Dec 2021 16:35:56 -0800
+Subject: [PATCH] Add system prop for disabling config comment parsing
```
Comment (jpenilla) [inline patches/api/0352-Add-system-prop-for-disabling-config-comment-parsing.patch]:
```suggestion
Subject: [PATCH] Add system prop for disabling config comment parsing by default

Allows for certain legacy plugins to continue to work without changes by setting `Paper.parseYamlCommentsByDefault` to false
```

**Review by jpenilla — CHANGES_REQUESTED** [review]

File: patches/api/0352-Add-system-prop-for-disabling-config-comment-parsing.patch (hunk 1 above)
Comment (Machine-Maker) [reply patches/api/0352-Add-system-prop-for-disabling-config-comment-parsing.patch]: Done, didnt do exactly that commit msg cause I don't like when the patch file cuts off abruptly, but I think it gets the msg across

**Review by jpenilla — APPROVED** [review]

**Pull request merged**